File: apps/web/src/lib/character-consistency.ts
import type {
  CharacterContinuityState,
  CharacterConsistencySelection,
  CharacterProfile,
  ComicPanel,
  RevisionIntent,
  PanelContinuitySnapshot,
  PanelConsistencyPlan
} from "@archmanga/shared";

export interface ConsistencyPreflight {
  status: "ready" | "caution" | "blocked";
  title: string;
  reasons: string[];
}

export interface ContinuityLockSuggestion {
  summary: string;
  promptSeed: string;
  sceneSummary: string;
  shotType: string;
  styleNotes: string;
  revisionIntent: RevisionIntent;
  carriedStates: CharacterContinuityState[];
}

function resolveContinuityCharacterIds(panel: ComicPanel, previousPanel?: ComicPanel) {
  return panel.characterIds.length ? panel.characterIds : previousPanel?.characterIds ?? [];
}

function resolveReadiness(score: number): PanelConsistencyPlan["readiness"] {
  if (score >= 75) {
    return "strong";
  }
  if (score >= 45) {
    return "partial";
  }
  return "weak";
}

function getPreviousCharacterState(
  previousPanel: ComicPanel | undefined,
  characterId: string
) {
  return previousPanel?.continuitySnapshot?.characterStates.find(
    (state) => state.characterId === characterId
  );
}

function selectPreferredRoles(panel: ComicPanel) {
  const shotType = panel.prompt.shotType.toLowerCase();
  const editPriority = panel.prompt.revisionIntent.editPriority;
  if (["close", "reaction", "portrait"].some((token) => shotType.includes(token)) || editPriority === "expression") {
    return ["primary", "face", "expression", "support", "full-body", "outfit"] as const;
  }
  if (["wide", "long", "full"].some((token) => shotType.includes(token)) || editPriority === "pose") {
    return ["full-body", "outfit", "primary", "support", "face", "expression"] as const;
  }
  return ["primary", "face", "full-body", "outfit", "expression", "support"] as const;
}

function scoreCharacterConsistency(panel: ComicPanel, character: CharacterProfile, activeReferenceCount: number) {
  let score = 0;
  if (character.appearance) {
    score += 18;
  }
  if (character.wardrobe) {
    score += 14;
  }
  if (character.referenceNotes) {
    score += 8;
  }
  if (character.consistency.anchorFeatures.length) {
    score += 18;
  }
  if (character.consistency.forbiddenDrift.length) {
    score += 10;
  }
  if (character.consistency.expressionDefaults.length) {
    score += 6;
  }
  if (activeReferenceCount) {
    score += 16;
  }
  if (character.adapter.enabled && activeReferenceCount) {
    score += 8;
  }
  if (panel.prompt.revisionIntent.preserveCharacterIdentity) {
    score += 6;
  }
  return Math.min(score, 100);
}

function selectReferenceImages(
  character: CharacterProfile,
  preferredRoles: readonly CharacterProfile["references"][number]["role"][],
  previousState?: CharacterContinuityState
) {
  const roleRank = new Map(preferredRoles.map((role, index) => [role, preferredRoles.length - index]));
  const adapterRank = new Map(
    character.adapter.referenceImageIds.map((referenceId, index) => [
      referenceId,
      character.adapter.referenceImageIds.length - index
    ])
  );
  const continuityRank = new Map(
    (previousState?.carriedReferenceIds ?? []).map((referenceId, index, ids) => [
      referenceId,
      ids.length - index
    ])
  );

  return [...character.references]
    .sort((left, right) => {
      const leftScore =
        (roleRank.get(left.role) ?? 0) * 100 +
        (continuityRank.get(left.id) ?? 0) * 350 +
        (adapterRank.get(left.id) ?? 0) * 10 +
        (left.role === "primary" ? 1 : 0);
      const rightScore =
        (roleRank.get(right.role) ?? 0) * 100 +
        (continuityRank.get(right.id) ?? 0) * 350 +
        (adapterRank.get(right.id) ?? 0) * 10 +
        (right.role === "primary" ? 1 : 0);
      return rightScore - leftScore;
    })
    .slice(0, 2);
}

function buildWarnings(
  panel: ComicPanel,
  character: CharacterProfile,
  preferredRoles: readonly CharacterProfile["references"][number]["role"][],
  selectedReferences: CharacterProfile["references"],
  previousState?: CharacterContinuityState
) {
  const roles = new Set(character.references.map((reference) => reference.role));
  const warnings: string[] = [];
  if (!character.references.length) {
    warnings.push("No reference images uploaded yet.");
  }
  if (character.adapter.enabled && !character.adapter.referenceImageIds.length) {
    warnings.push("Adapter is enabled but no adapter references are selected.");
  }
  if (
    ["primary", "face", "expression"].includes(preferredRoles[0]) &&
    !["primary", "face", "expression"].some((role) => roles.has(role as CharacterProfile["references"][number]["role"]))
  ) {
    warnings.push("Close-up identity references are missing for this shot.");
  }
  if (
    ["full-body", "outfit"].includes(preferredRoles[0]) &&
    !["full-body", "outfit"].some((role) => roles.has(role as CharacterProfile["references"][number]["role"]))
  ) {
    warnings.push("Full-body or outfit anchors are missing for this shot.");
  }
  if (panel.prompt.revisionIntent.preserveCharacterIdentity && !selectedReferences.length) {
    warnings.push("Identity lock is enabled, but this panel has no active reference anchors.");
  }
  if (
    previousState?.carriedReferenceIds.length &&
    !selectedReferences.some((reference) => previousState.carriedReferenceIds.includes(reference.id))
  ) {
    warnings.push("Previous-panel continuity anchors are available but not currently selected.");
  }
  return warnings;
}

export function buildPanelConsistencyPlan(
  panel: ComicPanel,
  characters: CharacterProfile[],
  previousPanel?: ComicPanel
): PanelConsistencyPlan {
  const characterPlans: CharacterConsistencySelection[] = characters.map((character) => {
    const preferredRoles = selectPreferredRoles(panel);
    const previousState = getPreviousCharacterState(previousPanel, character.id);
    const selectedReferences = selectReferenceImages(character, preferredRoles, previousState);
    const score = Math.min(
      100,
      scoreCharacterConsistency(panel, character, selectedReferences.length) +
        (previousState ? 8 : 0) +
        (selectedReferences.some((reference) => previousState?.carriedReferenceIds.includes(reference.id)) ? 4 : 0)
    );
    return {
      characterId: character.id,
      characterName: character.name,
      readiness: resolveReadiness(score),
      score,
      anchorSummary: [
        character.consistency.anchorFeatures.join(", "),
        character.consistency.bodyShape,
        character.appearance
      ]
        .filter(Boolean)
        .join("; "),
      wardrobeLock: previousState?.wardrobe || character.wardrobe,
      expressionCue:
        previousState?.expression ||
        character.consistency.expressionDefaults.join(", "),
      selectedReferenceIds: selectedReferences.map((reference) => reference.id),
      selectedReferenceLabels: selectedReferences.map((reference) => reference.label),
      selectedReferenceUrls: selectedReferences.map((reference) => reference.url),
      adapterProvider: character.adapter.provider,
      adapterEnabled: character.adapter.enabled,
      adapterWeight: character.adapter.weight,
      promptHints: [
        character.consistency.anchorFeatures.length
          ? `${character.name} anchor features: ${character.consistency.anchorFeatures.join(", ")}.`
          : `${character.name} appearance lock: ${character.appearance}.`,
        character.wardrobe ? `${character.name} wardrobe lock: ${character.wardrobe}.` : "",
        character.consistency.expressionDefaults.length
          ? `${character.name} expression baseline: ${character.consistency.expressionDefaults.join(", ")}.`
          : "",
        previousState?.expression
          ? `${character.name} carry forward expression: ${previousState.expression}.`
          : "",
        previousState?.wardrobe
          ? `${character.name} carry forward wardrobe: ${previousState.wardrobe}.`
          : "",
        previousState?.framingCue
          ? `${character.name} previous framing cue: ${previousState.framingCue}.`
          : "",
        previousState?.poseCue
          ? `${character.name} previous pose cue: ${previousState.poseCue}.`
          : "",
        selectedReferences.length
          ? `${character.name} selected references: ${selectedReferences.map((reference) => reference.label).join(", ")}.`
          : ""
      ].filter(Boolean),
      negativeHints: [
        character.consistency.forbiddenDrift.length
          ? `Avoid ${character.name} drift: ${character.consistency.forbiddenDrift.join(", ")}.`
          : "",
        character.negativePrompt
      ].filter(Boolean),
      warnings: buildWarnings(panel, character, preferredRoles, selectedReferences, previousState)
    };
  });

  const score = characterPlans.length
    ? Math.round(characterPlans.reduce((sum, plan) => sum + plan.score, 0) / characterPlans.length)
    : 0;
  const readiness = resolveReadiness(score);
  const referenceCount = characterPlans.reduce(
    (sum, plan) => sum + plan.selectedReferenceIds.length,
    0
  );
  return {
    readiness,
    score,
    summary: characterPlans.length
      ? `${readiness.toUpperCase()} consistency for ${characterPlans
          .map((plan) => plan.characterName)
          .join(", ")} with ${referenceCount} active reference anchors.`
      : "No active character anchors for this panel yet.",
    globalHints: [
      characterPlans.length ? "Keep wardrobe and silhouette locked across adjacent panels." : "",
      previousPanel?.continuitySnapshot?.characterStates.length
        ? "Carry forward previous-panel expression, wardrobe, and framing when the same characters remain in scene."
        : "",
      panel.prompt.revisionIntent.preserveCharacterIdentity
        ? "Prioritize face identity over style drift during redraws."
        : "",
      referenceCount
        ? "Use shot-matched references instead of generic anchors whenever possible."
        : "Add shot-matched references to improve identity stability.",
      characterPlans.length
        ? "Do not change costume, eye color, or hair silhouette unless explicitly requested."
        : ""
    ].filter(Boolean),
    characterPlans
  };
}

export function buildConsistencyPreflight(plan: PanelConsistencyPlan): ConsistencyPreflight {
  const reasons = plan.characterPlans.flatMap((characterPlan) => characterPlan.warnings);
  const hasMissingReferences = reasons.some((reason) =>
    ["No reference images", "no active reference anchors", "no adapter references"].some((token) =>
      reason.toLowerCase().includes(token.toLowerCase())
    )
  );
  if (plan.characterPlans.length > 0 && hasMissingReferences) {
    return {
      status: "blocked",
      title: "Consistency anchors need attention",
      reasons
    };
  }
  if (plan.readiness !== "strong" || reasons.length) {
    return {
      status: "caution",
      title: "Consistency is usable but not locked yet",
      reasons
    };
  }
  return {
    status: "ready",
    title: "Consistency anchors look strong",
    reasons: plan.globalHints
  };
}

export function buildContinuityLockSuggestion(
  panel: ComicPanel,
  previousPanel: ComicPanel | undefined,
  characters: CharacterProfile[]
): ContinuityLockSuggestion | null {
  const previousSnapshot = previousPanel?.continuitySnapshot;
  if (!previousSnapshot) {
    return null;
  }

  const continuityCharacterIds = resolveContinuityCharacterIds(panel, previousPanel);
  const carriedStates = previousSnapshot.characterStates.filter((state) =>
    continuityCharacterIds.includes(state.characterId)
  );
  if (!carriedStates.length) {
    return null;
  }

  const stateSummary = carriedStates
    .map((state) => {
      const bits = [
        state.characterName,
        state.expression ? `expression ${state.expression}` : "",
        state.wardrobe ? `wardrobe ${state.wardrobe}` : "",
        state.framingCue ? `framing ${state.framingCue}` : ""
      ].filter(Boolean);
      return bits.join(", ");
    })
    .join(" | ");

  const styleNotes = Array.from(
    new Set([panel.prompt.styleNotes, previousSnapshot.styleNotes].filter(Boolean))
  ).join(", ");
  const shotType = panel.prompt.shotType || previousSnapshot.shotType || carriedStates[0]?.framingCue || "";
  const sceneSummary =
    panel.prompt.sceneSummary || previousSnapshot.sceneSummary || "Continue the same scene state.";
  const promptSeed = panel.prompt.prompt.trim()
    ? panel.prompt.prompt
    : `Continue the previous beat while keeping ${stateSummary}.`;

  return {
    summary: `Carry forward ${stateSummary}.`,
    promptSeed,
    sceneSummary,
    shotType,
    styleNotes,
    revisionIntent: {
      ...panel.prompt.revisionIntent,
      preserveCharacterIdentity: true,
      lockCharacterAppearance: true,
      lockCharacterWardrobe: carriedStates.some((state) => Boolean(state.wardrobe)),
      lockCharacterExpression: carriedStates.some((state) => Boolean(state.expression)),
      lockCameraFraming: carriedStates.some((state) => Boolean(state.framingCue)),
      preserveBackground:
        panel.prompt.revisionIntent.preserveBackground || Boolean(previousSnapshot.sceneSummary),
      preserveComposition:
        panel.prompt.revisionIntent.preserveComposition ||
        carriedStates.some((state) => Boolean(state.framingCue && state.framingCue === shotType))
    },
    carriedStates
  };
}

export function applyContinuityDefaultsToPanel(
  panel: ComicPanel,
  previousPanel: ComicPanel | undefined,
  characters: CharacterProfile[]
): ComicPanel {
  if (!previousPanel) {
    return panel;
  }

  const continuityCharacterIds = resolveContinuityCharacterIds(panel, previousPanel);
  const suggestion = buildContinuityLockSuggestion(
    {
      ...panel,
      characterIds: continuityCharacterIds
    },
    previousPanel,
    characters
  );

  return {
    ...panel,
    mode: panel.mode || previousPanel.mode,
    modelId: panel.modelId || previousPanel.modelId,
    workflowPresetId: panel.workflowPresetId || previousPanel.workflowPresetId,
    characterIds: continuityCharacterIds,
    sceneMemoryId: panel.sceneMemoryId ?? previousPanel.sceneMemoryId,
    prompt: {
      ...panel.prompt,
      prompt: panel.prompt.prompt || suggestion?.promptSeed || "",
      sceneSummary:
        panel.prompt.sceneSummary || suggestion?.sceneSummary || previousPanel.prompt.sceneSummary,
      shotType: panel.prompt.shotType || suggestion?.shotType || previousPanel.prompt.shotType,
      styleNotes:
        panel.prompt.styleNotes || suggestion?.styleNotes || previousPanel.prompt.styleNotes,
      revisionIntent: suggestion
        ? {
            ...panel.prompt.revisionIntent,
            ...suggestion.revisionIntent
          }
        : panel.prompt.revisionIntent
    }
  };
}

export function buildPanelContinuitySnapshot(
  panel: ComicPanel,
  characters: CharacterProfile[],
  plan: PanelConsistencyPlan = buildPanelConsistencyPlan(panel, characters)
): PanelContinuitySnapshot {
  const characterStates: CharacterContinuityState[] = plan.characterPlans.map((characterPlan) => ({
    characterId: characterPlan.characterId,
    characterName: characterPlan.characterName,
    expression:
      characterPlan.expressionCue ||
      characters.find((character) => character.id === characterPlan.characterId)?.consistency.expressionDefaults[0] ||
      "",
    wardrobe: characterPlan.wardrobeLock,
    poseCue:
      panel.prompt.revisionIntent.editPriority === "pose"
        ? panel.prompt.revisionIntent.changeInstructions
        : panel.prompt.prompt,
    framingCue: panel.prompt.shotType,
    carriedReferenceIds: characterPlan.selectedReferenceIds,
    notes: [characterPlan.anchorSummary, ...characterPlan.promptHints.slice(0, 1)].filter(Boolean).join(" | ")
  }));

  return {
    continuitySummary: [
      panel.prompt.sceneSummary || "Continue the same scene state.",
      panel.prompt.shotType ? `Shot: ${panel.prompt.shotType}` : "",
      characterStates.length
        ? `Carry ${characterStates.map((state) => `${state.characterName}(${state.expression || "neutral"})`).join(", ")}`
        : "",
      plan.summary
    ]
      .filter(Boolean)
      .join(" | "),
    sourcePrompt: panel.prompt.prompt,
    shotType: panel.prompt.shotType,
    sceneSummary: panel.prompt.sceneSummary,
    styleNotes: panel.prompt.styleNotes,
    characterStates
  };
}

File: apps/api/app/services/character_consistency.py
from app.schemas.comic import (
    CharacterConsistencySelection,
    CharacterProfile,
    ComicPanel,
    PanelConsistencyPlan,
)


def build_panel_consistency_plan(
    panel: ComicPanel,
    characters: list[CharacterProfile],
    previous_panel: ComicPanel | None = None,
) -> PanelConsistencyPlan:
    character_plans = [
        build_character_consistency_selection(panel, character, previous_panel) for character in characters
    ]
    score = round(sum(item.score for item in character_plans) / len(character_plans)) if character_plans else 0
    readiness = _resolve_readiness(score)
    active_names = ", ".join(item.character_name for item in character_plans) if character_plans else "no active characters"
    reference_count = sum(len(item.selected_reference_ids) for item in character_plans)
    global_hints = [
        "Keep wardrobe and silhouette locked across adjacent panels." if character_plans else "",
        (
            "Carry forward previous-panel expression, wardrobe, and framing when the same characters remain in scene."
            if previous_panel
            and previous_panel.continuity_snapshot
            and previous_panel.continuity_snapshot.character_states
            else ""
        ),
        "Prioritize face identity over style drift during redraws."
        if panel.prompt.revision_intent.preserve_character_identity
        else "",
        "Use shot-matched references instead of generic anchors whenever possible."
        if reference_count
        else "Add shot-matched references to improve identity stability.",
        "Do not change costume, eye color, or hair silhouette unless the director explicitly asks for it."
        if character_plans
        else "",
    ]
    summary = (
        f"{readiness.title()} consistency plan for {active_names} with {reference_count} active reference "
        f"{'anchor' if reference_count == 1 else 'anchors'}."
        if character_plans
        else "No active character anchors for this panel yet."
    )
    return PanelConsistencyPlan(
        readiness=readiness,
        score=score,
        summary=summary,
        global_hints=[hint for hint in global_hints if hint],
        character_plans=character_plans,
    )


def build_character_consistency_selection(
    panel: ComicPanel,
    character: CharacterProfile,
    previous_panel: ComicPanel | None = None,
) -> CharacterConsistencySelection:
    effective_revision_intent = _build_character_revision_intent(panel, character.id)
    character_lock_override = _get_character_lock_override(panel, character.id)
    preferred_roles = _select_preferred_roles(panel)
    previous_state = _get_previous_character_state(previous_panel, character.id)
    references = _select_reference_images(character, preferred_roles, previous_state)
    selected_reference_ids = [reference.id for reference in references]
    selected_reference_labels = [reference.label for reference in references]
    selected_reference_urls = [reference.url for reference in references]
    prompt_hints = [
        f"{character.name} anchor features: {', '.join(character.consistency.anchor_features)}."
        if character.consistency.anchor_features
        else f"{character.name} appearance lock: {character.appearance}.",
        f"{character.name} wardrobe lock: {character.wardrobe}." if character.wardrobe else "",
        (
            f"{character.name} expression baseline: {', '.join(character.consistency.expression_defaults)}."
            if character.consistency.expression_defaults
            else ""
        ),
        f"{character.name} appearance continuity is explicitly locked for this panel."
        if effective_revision_intent.lock_character_appearance
        else "",
        f"{character.name} wardrobe continuity is explicitly locked for this panel."
        if effective_revision_intent.lock_character_wardrobe
        else "",
        f"{character.name} expression continuity is explicitly locked for this panel."
        if effective_revision_intent.lock_character_expression
        else "",
        f"{character.name} framing continuity is explicitly locked for this panel."
        if effective_revision_intent.lock_camera_framing
        else "",
        f"{character.name} per-character lock override: {_summarize_character_lock(character_lock_override)}."
        if character_lock_override and _summarize_character_lock(character_lock_override)
        else "",
        f"{character.name} carry forward expression: {previous_state.expression}."
        if previous_state and previous_state.expression
        else "",
        f"{character.name} carry forward wardrobe: {previous_state.wardrobe}."
        if previous_state and previous_state.wardrobe
        else "",
        f"{character.name} previous framing cue: {previous_state.framing_cue}."
        if previous_state and previous_state.framing_cue
        else "",
        f"{character.name} previous pose cue: {previous_state.pose_cue}."
        if previous_state and previous_state.pose_cue
        else "",
        (
            f"{character.name} selected references: {', '.join(selected_reference_labels)}."
            if selected_reference_labels
            else ""
        ),
    ]
    negative_hints = [
        f"Avoid {character.name} drift: {', '.join(character.consistency.forbidden_drift)}."
        if character.consistency.forbidden_drift
        else "",
        character.negative_prompt,
    ]
    warnings = _build_warnings(
        panel,
        character,
        preferred_roles,
        references,
        previous_state,
        effective_revision_intent,
    )
    score = min(
        100,
        _score_character_consistency(character, references, effective_revision_intent)
        + (8 if previous_state else 0)
        + (
            4
            if previous_state
            and any(reference.id in previous_state.carried_reference_ids for reference in references)
            else 0
        ),
    )
    return CharacterConsistencySelection(
        character_id=character.id,
        character_name=character.name,
        readiness=_resolve_readiness(score),
        score=score,
        anchor_summary=_join_non_empty(
            [", ".join(character.consistency.anchor_features), character.consistency.body_shape, character.appearance]
        ),
        wardrobe_lock=previous_state.wardrobe if previous_state and previous_state.wardrobe else character.wardrobe,
        expression_cue=(
            previous_state.expression
            if previous_state and previous_state.expression
            else ", ".join(character.consistency.expression_defaults)
        ),
        selected_reference_ids=selected_reference_ids,
        selected_reference_labels=selected_reference_labels,
        selected_reference_urls=selected_reference_urls,
        adapter_provider=character.adapter.provider,
        adapter_enabled=character.adapter.enabled,
        adapter_weight=character.adapter.weight,
        prompt_hints=[hint for hint in prompt_hints if hint],
        negative_hints=[hint for hint in negative_hints if hint],
        warnings=warnings,
    )


def _select_preferred_roles(panel: ComicPanel) -> list[str]:
    shot_type = panel.prompt.shot_type.lower()
    edit_priority = panel.prompt.revision_intent.edit_priority
    if any(token in shot_type for token in ["close", "reaction", "portrait"]) or edit_priority == "expression":
        return ["primary", "face", "expression", "support", "full-body", "outfit"]
    if any(token in shot_type for token in ["wide", "long", "full"]) or edit_priority == "pose":
        return ["full-body", "outfit", "primary", "support", "face", "expression"]
    return ["primary", "face", "full-body", "outfit", "expression", "support"]


def _select_reference_images(character: CharacterProfile, preferred_roles: list[str], previous_state=None):
    role_rank = {role: len(preferred_roles) - index for index, role in enumerate(preferred_roles)}
    adapter_rank = {
        reference_id: len(character.adapter.reference_image_ids) - index
        for index, reference_id in enumerate(character.adapter.reference_image_ids)
    }
    continuity_rank = {
        reference_id: len(previous_state.carried_reference_ids) - index
        for index, reference_id in enumerate(previous_state.carried_reference_ids)
    } if previous_state else {}
    ranked = sorted(
        character.references,
        key=lambda reference: (
            role_rank.get(reference.role, 0) * 100
            + continuity_rank.get(reference.id, 0) * 350
            + adapter_rank.get(reference.id, 0) * 10
            + (1 if reference.role == "primary" else 0)
        ),
        reverse=True,
    )
    return ranked[:2]


def _build_warnings(
    panel: ComicPanel,
    character: CharacterProfile,
    preferred_roles: list[str],
    references: list,
    previous_state=None,
    effective_revision_intent=None,
):
    warnings: list[str] = []
    roles = {reference.role for reference in character.references}
    if not character.references:
        warnings.append("No reference images uploaded yet.")
    if character.adapter.enabled and not character.adapter.reference_image_ids:
        warnings.append("Adapter is enabled but no adapter references are selected.")
    if preferred_roles[0] in {"primary", "face", "expression"} and not roles.intersection({"primary", "face", "expression"}):
        warnings.append("Close-up identity references are missing for this shot.")
    if preferred_roles[0] in {"full-body", "outfit"} and not roles.intersection({"full-body", "outfit"}):
        warnings.append("Full-body or outfit anchors are missing for this shot.")
    if effective_revision_intent and effective_revision_intent.preserve_character_identity and not references:
        warnings.append("Identity lock is enabled, but this panel has no active reference anchors.")
    if (
        previous_state
        and previous_state.carried_reference_ids
        and not any(reference.id in previous_state.carried_reference_ids for reference in references)
    ):
        warnings.append("Previous-panel continuity anchors are available but not currently selected.")
    return warnings


def _get_previous_character_state(previous_panel: ComicPanel | None, character_id: str):
    if not previous_panel or not previous_panel.continuity_snapshot:
        return None
    for state in previous_panel.continuity_snapshot.character_states:
        if state.character_id == character_id:
            return state
    return None


def _score_character_consistency(character: CharacterProfile, references: list, effective_revision_intent) -> int:
    score = 0
    if character.appearance:
        score += 18
    if character.wardrobe:
        score += 14
    if character.reference_notes:
        score += 8
    if character.consistency.anchor_features:
        score += 18
    if character.consistency.forbidden_drift:
        score += 10
    if character.consistency.expression_defaults:
        score += 6
    if references:
        score += 16
    if character.adapter.enabled and references:
        score += 8
    if effective_revision_intent.preserve_character_identity:
        score += 6
    return min(score, 100)


def _get_character_lock_override(panel: ComicPanel, character_id: str):
    for character_lock in panel.prompt.revision_intent.character_locks:
        if character_lock.character_id == character_id:
            return character_lock
    return None


def _build_character_revision_intent(panel: ComicPanel, character_id: str):
    revision_intent = panel.prompt.revision_intent
    character_lock = _get_character_lock_override(panel, character_id)
    if not character_lock:
        return revision_intent
    updated = revision_intent.model_copy(deep=True)
    if character_lock.preserve_character_identity is not None:
        updated.preserve_character_identity = character_lock.preserve_character_identity
    if character_lock.lock_character_appearance is not None:
        updated.lock_character_appearance = character_lock.lock_character_appearance
    if character_lock.lock_character_wardrobe is not None:
        updated.lock_character_wardrobe = character_lock.lock_character_wardrobe
    if character_lock.lock_character_expression is not None:
        updated.lock_character_expression = character_lock.lock_character_expression
    if character_lock.lock_camera_framing is not None:
        updated.lock_camera_framing = character_lock.lock_camera_framing
    return updated


def _summarize_character_lock(character_lock) -> str:
    if not character_lock:
        return ""
    bits = [
        "appearance locked" if character_lock.lock_character_appearance else "",
        "wardrobe locked" if character_lock.lock_character_wardrobe else "",
        "expression locked" if character_lock.lock_character_expression else "",
        "camera locked" if character_lock.lock_camera_framing else "",
        "identity protected" if character_lock.preserve_character_identity else "",
        character_lock.note,
    ]
    return ", ".join(bit for bit in bits if bit)


def _resolve_readiness(score: int) -> str:
    if score >= 75:
        return "strong"
    if score >= 45:
        return "partial"
    return "weak"


def _join_non_empty(parts: list[str]) -> str:
    return "; ".join(part for part in parts if part)
